CANTroller2/src/uictrl.h

```diff
@@ -1,9 +1,36 @@
 #ifndef UICTRL_H
 #define UICTRL_H
 
-#include "globals.h"
+#include "utils.h"
+#include <memory> // for std::shared_ptr
 #include "FunctionalInterrupt.h"
 
+// Potentiometer does an analog read from a pin and maps it to a percent (0%-100%). We filter the value to keep it smooth. The output can either be read as a value, or passed as a pointer so that other components can
+// access it directly as needed.
+class Potentiometer {
+    protected:
+        // static constexpr float adc_min = 300; // TUNED 230603 - Used only in determining theconversion factor
+        static constexpr float adc_min = 300; // TUNED 230603 - Used only in determining theconversion factor
+        static constexpr float adc_max = 4095; // TUNED 230613 - adc max measured = ?, or 9x.? % of adc_range. Used only in determining theconversion factor
+        static constexpr float _ema_alpha = 0.1;
+        uint8_t _pin;
+        std::shared_ptr<float> _val;
+    public:
+        static constexpr float min = 0.0; // percent min
+        static constexpr float max = 100.0; // percent max
+        Potentiometer(uint8_t arg_pin) : _pin(arg_pin), _val(std::make_shared<float>(0.0)) {}
+        Potentiometer() = delete; // must have a pin defined
+        void setup() {
+            set_pin(_pin, INPUT);
+        }
+        void update() {
+            float new_val = constrain(map(static_cast<float>(analogRead(_pin)), adc_min, adc_max, min, max), min, max);

```

Comment:
constrain() is redundant, as map already constrains between min and max, right?

Reply:
yep, removed

Reply:
ah actually, now i remember why this is here. the lower  bound of the pin reading for the pot varies (mine is ~28 but yours might be different, the value soren has here is 300). if we map it from the theoretical adc min of 0, we never quite get to zero, and the filtered value is so small that it bounces around at a fractional value forever. If we use a higher bound we can get closer to zero, but it's hard to find one that works consistently, so instead we set a bound that's higher than we'll ever get (which causes us to have negative values due to the mapping) and then constrain it (so values below zero go to zero). This means that we have a deadband at the bottom of the range where everything is zero, and we map the rest of the range accordingly. It's maybe not the slickest solution but it makes the zeropoint solid and not jumpy.

I'll add a comment so it's more obvious.